File: DSA/DSA11026 DUYỆT SAU CÂY TÌM KIẾM CÂN BẰNG.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};


void add(TreeNode** root, int node)
{
    if (!(*root))
    {
        (*root) = new TreeNode(node);
    }
    else if (node < (*root)->val)
    {
        if (!(*root)->left) (*root)->left = new TreeNode(node);
        else add(&(*root)->left, node);
    }
    else if (node > (*root)->val)
    {
        if (!(*root)->right) (*root)->right = new TreeNode(node);
        else add(&(*root)->right, node);
    }
}

void insert(TreeNode** root, vector<int> node)
{
    if (node.empty())
    {
        return;
    }
    
    int mid = (node.size()-1)/2;

    add(&(*root), node[mid]);

    if (node.size() > 1)
    {
        vector<int> left(node.begin(), node.begin()+mid);
        vector<int> right(node.begin()+mid+1, node.end());
        
        insert(root, left);
        insert(root, right);
    }
}

void post_order(TreeNode* node)
{
    if (!node)
    {
        return;
    }

    post_order(node->left);
    post_order(node->right);
    cout << node->val << " ";
}

void submit()
{
    int n;
    cin >> n;
    vector<int> node(n);
    for (auto &&i : node) cin >> i;

    sort(node.begin(), node.end());

    TreeNode* root = nullptr;

    insert(&root, node);

    post_order(root);

    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
} 

File: DSA_python/DSA11026 DUYỆT SAU CÂY TÌM KIẾM CÂN BẰNG.py
class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

def add_to_tree(root, node):
    if root.val > node:
        if not root.left: root.left = TreeNode(node)
        else: add_to_tree(root.left, node)
    elif root.val < node:
        if not root.right: root.right = TreeNode(node)
        else: add_to_tree(root.right, node)

def change_to_tree(root, edges):
    if len(edges) == 0: return

    mid = (len(edges)-1)//2

    add_to_tree(root, edges[mid])

    if len(edges) > 1:
        left = edges[:mid]
        right = edges[mid+1:]

        change_to_tree(root, left)
        change_to_tree(root, right)

def post_order(node):
    if not node:
        return

    post_order(node.left)
    post_order(node.right)
    print(node.val, end=" ")

def submit():
    n = int(input())
    edges = list(map(int, input().split()))
    edges.sort()

    root = TreeNode(edges[(n-1)//2])
    change_to_tree(root, edges)

    post_order(root)

    print()











test = int(input())
for i in range(test):
    submit()